enhancement_proposals/sbep_1.md

```diff
@@ -0,0 +1,172 @@
+
+`BaseEstimator` ties together the different algorithm categories through a
+common high-level interface for learning (fitting) parameters from data by inheriting
+from `BaseObject` and providing an additional interface for *fittable* learning
+algorithms, including:
+
+- An instance `is_fitted` property denoting whether the the estimator has been
+  fit (`is_fitted` does this by inspecting a non-public `_is_fitted` attribute
+  set in each algorithm's call to its `fit` method).
+- An instance `check_is_fitted` method for raising an error when an estimator
+  has not been fitted yet.
+
+Although the `BaseEstimator` interface may seem like it should include a `fit`
+method for learning the parameters from data, it is not included. Instead
+`BaseEstimator` assumes sub-classes implement a `fit` method (and that this
+method appropriately sets the `_is_fitted` attribute), since the signature of
+`fit` is learning task specific. Therefore, the specific `fit` implementation
+is left to child classes implemented outside of `skbase`.
+
+#### HeterogenousMetaObject
+
+**DESIGN DECISION: Do we want to provide specific base classes for useful
+implementations of HeterogenousMetaEstimator? Specifically, base pipeline
+classes (maybe a standard "linear" step based pipeline and eventually a
+DAG based pipeline?). In both cases, we'd be defining interface and some underlying
+functionality, but not the actual methods to execute the pipelines. For example,
+we might have `steps` parameter in constructor and basic validation of format
+that users shoudl provide steps in (and verify that input contains interface
+compliant objects). But we may decide this is beyond the scope of `skbase`.**
+
+### Tools For Working With Base Classes
+
+`skbase` should make it easy for developers to work with BaseObjects and create
+their own packages that follow `skbase`'s principles. To accomplish this
+`skbase` includes tools that make it easier to accomplish common workflows
+that arise.
+
+#### `skbase.lookup`: Collecting (Retrieving) Information on BaseObjects and Package Metadata
+
+The need to lookup classes arises in several contexts when working working with
+parametric objects, including the need to collect *similar* objects for
+testing or reporting.
+
+`skbase` provides this through two function interfaces:
+
+- `all_objects` provides the ability to recursively walk packages (or sub-packages)
+  and return the objects that meet the filters specified in the function call.
+- `get_package_metadata` provides the ability to recursively walk a package's moduels
+  and collect information on the items contained in the modules (including
+  objects and functions)
+
+#### `skbase.testing`: Testing BaseObjects
+TO BE ADDED.
+
+#### `skbase.validate`: Validating and Comparing BaseObjects
+
+### Example Repository
+This would be a simple example package that illustrates how `skbase`'s functionality
+can be used to create another package. It will also be used by us to test
```

Comment:
hm, but it would not be the only way of testing it, right?

Reply:
Nope, good point on that. But it would provide us some testing of a separate repository.